yunwoo/programmers-1/가장-가까운-같은-글자.js

```diff
@@ -0,0 +1,26 @@
+function solution(s) {
+  const result = [];
+  const charArray = [];
+  const reverseString = s.split('').reverse();
+
+  for (let i = 0; i < s.length; i++) {
+    const pop = reverseString.pop();
+    const closeDistance =
+      charArray.lastIndexOf(pop) === -1
+        ? -1
+        : charArray.length - charArray.lastIndexOf(pop);
+
+    result.push(closeDistance);
+    charArray.push(pop);
+  }
+
+  return result;
+}
```

Comment:
```suggestion
function solution(s) {
  const result = [];
  const charArray = [];
  const reverseString = s.split('').reverse();

  for (let i = 0; i < s.length; i++) {
    const pop = reverseString.pop();
    const lastIdx = charArray.lastIndexOf(pop);
    const closeDistance = lastIdx === -1 ? -1 : charArray.length - lastIdx;

    result.push(closeDistance);
    charArray.push(pop);
  }

  return result;
}
```
- 풀이 잘봤습니다~ 🙇‍♂️
- charArray.lastIndexOf(pop)가 반복적으로 사용되기 때문에 연속으로 연산되지않게 변수로 빼서 1회만 호출해서 사용하는게 좀 더 효율적일 것 같습니다~
- 아래 코드는 split, map으로 O(N)으로 푼 풀이입니다. 현재 코드에서는 lastIndexOf는 O(N) 메서드라 최악의 경우 O(N^2)의 시간 복잡도를 갖고있어요!
```js
function solution(s) {
  const lastPositionObj = {};

  return s.split("").map((char, i) => {
    const distance = lastPositionObj[char] ? i + 1 - lastPositionObj[char] : -1;

    lastPositionObj[char] = lastPositionObj[char]
      ? Math.max(lastPositionObj[char], i + 1)
      : i + 1;
    return distance;
  });
}
```

Reply:
- shift를 피하는게 중요한게 아니였군요..😅
- 역시 객체.. 다음엔 꼭 객체를 활용해 보겠습니다ㅎㅎ 감사합니다🙇‍♂️